plugins/webp-uploads/picture-element.php

```diff
@@ -0,0 +1,154 @@
+	 *
+	 * The mime types will output in the picture element in the order they are provided.
+	 * The original image will be used as the fallback image for browsers that don't support the picture element.
+	 *
+	 * @since n.e.x.t
+	 * @param array<string> $mime_types    Mime types than can be used.
+	 * @param int           $attachment_id The id of the image being evaluated.
+	 */
+	$enabled_mime_types = apply_filters(
+		'webp_uploads_picture_element_mime_types',
+		array(
+			'image/webp',
+			'image/jpeg',
+		),
+		$attachment_id
+	);
+
+	$mime_types = array_filter(
+		$enabled_mime_types,
+		static function ( $mime_type ) use ( $sub_size_mime_types ) {
+			return in_array( $mime_type, $sub_size_mime_types, true );
+		}
+	);
+
+	// No eligible mime types.
+	if ( ! $mime_types ) {
+		return $image;
+	}
+
+	// If the original mime types is the only one available, no picture element is needed.
+	if ( 1 === count( $mime_types ) && $mime_types[0] === $original_file_mime_type ) {
+		return $image;
+	}
+
+	// Add each mime type to the picture's sources.
+	$picture_sources = '';
+
+	// Extract sizes using regex to parse image tag, then use to retrieve tag.
+	$image_sizes = array();
+	preg_match( '/width="([^"]+)"/', $image, $image_sizes );
+	$width = isset( $image_sizes[1] ) ? (int) $image_sizes[1] : false;
+	preg_match( '/height="([^"]+)"/', $image, $image_sizes );
+	$height      = isset( $image_sizes[1] ) ? (int) $image_sizes[1] : false;
+	$size_to_use = ( $width && $height ) ? array( $width, $height ) : 'full';
+
+	$image_src = wp_get_attachment_image_src( $attachment_id, $size_to_use, false );
+	if ( ! $image_src ) {
+		return $image;
+	}
+	list( $src, $width, $height ) = $image_src;
+	$size_array                   = array( absint( $width ), absint( $height ) );
+
+	foreach ( $mime_types as $image_mime_type ) {
+		$sizes = wp_calculate_image_sizes( $size_array, $src, $image_meta, $attachment_id );
+		// Filter core's wp_get_attachment_image_srcset to return the sources for the current mime type.
+
+		add_filter(
+			'wp_calculate_image_srcset',
+			static function ( $sources ) use ( $mime_type_data, $image_mime_type ) {
+				$filtered_sources = array();
+				foreach ( $sources as $source ) {
+					// Swap the URL for the current mime type.
+					if ( isset( $mime_type_data[ $image_mime_type ][ $source['descriptor'] ][ $source['value'] ] ) ) {
+						$filename  = $mime_type_data[ $image_mime_type ][ $source['descriptor'] ][ $source['value'] ]['file'];
+						$url_array = explode( '/', $source['url'] );
+						array_pop( $url_array );
+
+						$filtered_sources[] = array(
+							'url'        => implode( '/', $url_array ) . '/' . $filename,
+							'descriptor' => $source['descriptor'],
+							'value'      => $source['value'],
+						);
+					}
+				}
+				return $filtered_sources;
+			}
+		);
+		$image_srcset = wp_get_attachment_image_srcset( $attachment_id, $size_array, $image_meta );
+		remove_all_filters( 'wp_calculate_image_srcset' );
```

Comment:
The `remove_all_filters( 'wp_calculate_image_srcset' )` call could remove filters that other plugins have added, so it should just remove the one filter:

```suggestion
		$filter = static function ( $sources ) use ( $mime_type_data, $image_mime_type ): array {
			$filtered_sources = array();
			foreach ( $sources as $source ) {
				// Swap the URL for the current mime type.
				if ( isset( $mime_type_data[ $image_mime_type ][ $source['descriptor'] ][ $source['value'] ] ) ) {
					$filename  = $mime_type_data[ $image_mime_type ][ $source['descriptor'] ][ $source['value'] ]['file'];
					$url_array = explode( '/', $source['url'] );
					array_pop( $url_array );

					$filtered_sources[] = array(
						'url'        => implode( '/', $url_array ) . '/' . $filename,
						'descriptor' => $source['descriptor'],
						'value'      => $source['value'],
					);
				}
			}
			return $filtered_sources;
		};
		add_filter( 'wp_calculate_image_srcset', $filter );
		$image_srcset = wp_get_attachment_image_srcset( $attachment_id, $size_array, $image_meta );
		remove_filter( 'wp_calculate_image_srcset', $filter );
```